internal/manager/controllers.go

```diff
@@ -0,0 +1,85 @@
+package manager
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"k8s.io/apimachinery/pkg/types"
+	ctlr "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	"github.com/nginxinc/nginx-kubernetes-gateway/internal/reconciler"
+)
+
+const (
+	// addIndexFieldTimeout is the timeout used for adding an Index Field to a cache.
+	addIndexFieldTimeout = 2 * time.Minute
+)
+
+type controllerConfig struct {
+	objectType           client.Object
+	namespacedNameFilter reconciler.NamespacedNameFilterFunc // optional
+	k8sEventFilter       predicate.Predicate                 // optional
+	fieldIndexes         map[string]client.IndexerFunc       // optional
+}
+
+// newReconciler creates a new Implementation. Used for unit testing.
+var newReconciler = reconciler.NewImplementation
```

Comment:
Why use a global variable here instead of dependency injection?

Reply:
the signature of registerController is already large enough, I didn't want to pollute it.

Reply:
I don't think we should use mutable global variables as a way to reduce the number of function parameters. The global variable is also only used for testing purposes. In addition, one of the primary use cases for dependency injection is that it makes testing easier...